### PR TITLE
feat: adds beta opt-in redirects for vault and waypoint

### PR DESCRIPTION
**Pull request by BRKalow**
## ✅ "Ready for Review" checklist

- [ ] The Vercel preview link has been added to this PR's description
- [ ] The Asana task has been added to this PR's description
- [ ] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [ ] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [ ] The description template has been filled in below

---

## 🔗 Relevant links

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- [Asana task](url) 🎟️

## 🗒️ What

- Adds beta redirects for waypoint and vault when the opt-in cookie exists.
- Adds opt-out support via the `betaOptOut` query param

### CODE REVIEW

File: build-libs/__tests__/build-beta-opt-in-redirect.test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            "value": "true",
+          },
+          Object {
+            "type": "host",
+            "value": "(www\\\\.waypointproject\\\\.io|test-wp\\\\.hashi-mktg\\\\.com)",
+          },
+        ],
+        "permanent": false,
+        "source": "/:base(docs|commands|plugins)/:path*",
+      }
```
Comment (BRKalow) [inline build-libs/__tests__/build-beta-opt-in-redirect.test.ts:23]: This is gonna be hard to test prior to being in prod 😬, but the definition looks right!

File: src/pages/_middleware.ts [hunk 2]
```diff
@@ -76,8 +51,22 @@ export function middleware(req: NextRequest, ev: NextFetchEvent) {
...
+   * clear it and redirect back to the current URL without the betaOptOut query param
+   */
+  if (
+    product in __config.dev_dot.beta_product_slugs &&
+    params.get('betaOptOut') === 'true'
+  ) {
+    const url = req.nextUrl.clone()
+    url.searchParams.delete('betaOptOut')
+    return NextResponse.redirect(url).clearCookie(`${product}-io-beta-opt-in`)
+  }
```
Comment (BRKalow) [inline src/pages/_middleware.ts:67]: I validated this by removing the product check on L61 and testing locally. Saw that the cookie gets deleted and the URL is correctly redirected 👍

File: build-libs/__tests__/build-beta-opt-in-redirect.test.ts (hunk 1 above)
Comment (BRKalow) [reply build-libs/__tests__/build-beta-opt-in-redirect.test.ts:23]: FWIW, I validated without the host condition locally. All good.

File: src/components/opt-in-out/index.tsx [hunk 3]
```diff
@@ -25,16 +26,16 @@ export const PLATFORM_OPTIONS: PlatformOptionRedirectData = {
...
     },
     cookieKey: 'waypoint-io-beta-opt-in',
     cookieAnalyticsKey: 'waypoint-io-beta-opt-in-tracked',
   },
   'vault-io': {
     base_url: 'https://www.vaultproject.io/',
-    getRedirectPath: function TODO() {
-      return 'test'
+    getRedirectPath(path) {
+      return this.base_url + getIoRedirectPath(path)
```
Comment (kendallstrautman) [inline src/components/opt-in-out/index.tsx:38]: We'll still need to [add the opt in component ](https://github.com/hashicorp/dev-portal/blob/main/src/views/product-tutorials-view/index.tsx#L76)to the docs layouts, not sure if you wanted to handle that on a different PR.

Comment (BRKalow) [reply src/components/opt-in-out/index.tsx:38]: Yep! Planning to add that in a follow-up 👍

Comment (kendallstrautman) [reply src/components/opt-in-out/index.tsx:38]: Oh just watched you vid, saw that you want to do that in a separate PR, please disregard 😄

File: build-libs/build-beta-opt-in-redirect.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * Construct a redirect definition for beta product opt-in, based on cookies and host conditions
+ *
+ * @param {*} product  string
+ * @param {*} basePaths string[]
+ * @returns {Redirect} redirect
+ */
+function buildBetaProductOptInRedirect(product, basePaths) {
+  return {
+    source: `/:base(${basePaths.join('|')})/:path*`,
+    destination: `${__config.dev_dot.canonical_base_url}/${product}/:base/:path*`,
```
Comment (kendallstrautman) [inline build-libs/build-beta-opt-in-redirect.js:17]: Great usecase for the generic path matching 🌟😄

**Review by kendallstrautman — APPROVED** [review]
Looks great! Is there a way to test the redirects on a .io branch preview?

**Comment by BRKalow** [comment]
@kendallstrautman I can push up a change to temporarily remove the host condition and look at the `io_preview` cookie instead. One minute!

**Comment by BRKalow** [comment]
Validated in preview that the redirect is applied when the `waypoint-io-beta-opt-in=true` cookie exists

**Pull request merged**